### PR TITLE
Add scalar weighting model and listener

### PR DESCRIPTION
**Pull request by shadeMe**
Field names in `PyTorchTransformerOutput`, `TransformerModelOutput` and `DocTransformerOutput` were additionally modified to reduce ambiguity.

### CODE REVIEW

**Comment by shadeMe** [comment]
If we hold off on merging until [this `thinc` PR](https://github.com/explosion/thinc/pull/796) gets merged into `master`, we can pin the `thinc` requirement directly on the merge commit.

File: curated_transformers/models/scalar_weight.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    def __init__(self, *, num_layers: int, dropout_prob: float = 0.1):
+        super().__init__()
+
+        self.layer_weights = torch.nn.parameter.Parameter(
+            torch.zeros((num_layers), requires_grad=True)
+        )
+        self.scale = torch.nn.parameter.Parameter(
+            torch.tensor((1.0,), requires_grad=True)
+        )
+        self.dropout = torch.nn.Dropout(p=dropout_prob)
```
Comment (danieldk) [inline curated_transformers/models/scalar_weight.py:30]: I think we can just store the probability, since the dropout layer itself is never used?

File: curated_transformers/pipe.py [hunk 2]
```diff
@@ -175,7 +192,108 @@ def backprop(dYs):
...
                 outputs.append(model.ops.alloc2f(len(doc), width))
             else:
-                output, _ = pooling(doc._.trf_data.last_hidden_state, is_train)
+                output, _ = pooling(doc._.trf_data.last_hidden_layer_state, is_train)
+                outputs.append(output)
+
+        return outputs, lambda dX: []
+
+
+class ScalarWeightingLayerListener(TransformerListener):
```
Comment (danieldk) [inline curated_transformers/pipe.py:201]: Maybe just `ScalarWeightingListener`?

File: curated_transformers/models/scalar_weight.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    ops = model.ops
+    layer_hidden_sizes = [x.data.shape[1] for x in X]
+    if not all_equal(layer_hidden_sizes):
+        raise ValueError(
+            f"Not all hidden sizes are equal in input passed to scalar weight"
+        )
+
+    Xt = []
+    for layer in X:
+        batch = ops.alloc3f(1, layer.data.shape[0], layer.data.shape[1])
```
Comment (danieldk) [inline curated_transformers/models/scalar_weight.py:117]: Is it necessary to have the initial singleton dimension? Maybe it's better to modify the PyTorch model to not require this dimension?

Another thing that I wondered about -- since we are copying here anyway, maybe we should allocate + fill a single tensor for all layers, so that the `stack` is not necessary in the PyTorch model?

Comment (shadeMe) [reply curated_transformers/models/scalar_weight.py:117]: I originally had it implemented in such a way that we'd pass the entire document batch as its input. But it turned out to be unwieldy as the pooling didn't support batching. I'll stack the tensors without the batch dim during input conversion.

**Review by danieldk — APPROVED** [review]
